content/posts/2023-07-24-breakdown-of-http-clients-in-elixir/index.md

```diff
@@ -0,0 +1,458 @@
+
+Mint supports HTTP/1.1 and HTTP/2 out of the box, as well as WebSocket through [mint_web_socket].
+
+#### When to Use Mint
+
+Generally, ***don't use Mint***. Seriously. You know I mean this advice, because
+I'm one of the [two][gh-whatyouhide] [people][gh-ericmj] who maintain and
+originally created Mint itself! However, for most use cases, Mint is too low
+level. When you use it, you'll have to care about things such as pooling
+connections, process architecture, keeping the connection structs around, and so
+on. It's a bit like what you'd do in other cases, after all. For example, you're
+unlikely to use `:gen_tcp` to communicate directly with your PostgreSQL
+database. Instead, you'd probably reach *at least* for something like
+[Postgrex][postgrex] to abstract a lot of the complexity away.
+
+Still, there are some use cases where Mint can make a lot of sense. First and
+foremost, you can use it to build higher-level abstractions. That's exactly what
+a library called Finch does, which we'll talk about in a bit. Mint can also be
+useful in cases where you need fine-grained control over the performance and
+process architecture of your application. For example, say you have a fine-tuned
+[GenStage][gen_stage] pipeline where you need to make some HTTP calls at some
+point. GenStage stages are already processes, so having an HTTP client based on
+a process might introduce an unnecessary layer of processes in your application.
+Mint being processless solves exactly that.
+
+![Drawing of a GenStage pipeline with a traditional process-based HTTP client on
+the left, and the same pipeline but with Mint as the HTTP client on the
+right](./sketch-genstage.png)
+
+A few years ago, I worked at a company where we would've likely used Mint in
+exactly this way. At the time, I wrote [a blog
+post][forza-football-gen_stage-post] that goes into more detail in case you're
+interested.
+
+#### Bonus: Why Isn't Mint in the Elixir Standard Library?
+
+That's a great question! When we introduced Mint back in 2019, we [posted about
+it][mint-blog-post] on Elixir's website. Our original intention was to ship Mint
+with Elixir's standard library. This is also one of the reasons why we wrote
+Mint in Elixir, instead of Erlang. However, we then realized that it worked well
+as a standalone library, and including it into the standard library would
+increase the cost of maintaining the language as well as potentially slow down
+the development of Mint itself.
+
+That said, I think of Mint as the "standard-library HTTP client", that is, the
+low-level client that you'd expect in the standard library of a language like
+Elixir.
+
+### Finch
+
+[Finch][finch] is a client built on top of Mint. It serves an important job in
+the "pyramid of abstractions" of HTTP clients listed in this post: **pooling**.
+Finch provides pooling for Mint connections. Using Mint on its own means
+implementing some sort of strategy to store and pool connections, which is what
+Finch provides.
```

Comment:
It may be worth talking about HTTP 1 vs HTTP 2, how they are pooling vs multiplexing?